File: Chapter 04/C#/Section_4_3_Cross_Cutting_Concerns/After/OrderService.cs
namespace Chapter04.Section_4_3.After
{
    // 2. The Class "asks" for the dependency
    public class OrderService
    {
        private readonly ILogger _logger;

        // The dependency is "injected" via the constructor!
        public OrderService(ILogger logger)
        {
            _logger = logger;
        }

        public void SaveOrder(Order order)
        {
            // 3. Use the abstraction (follows DIP)
            _logger.Log("Saving order...");
            Console.WriteLine("(AFTER_SERVICE) Order saved.");
        }
    }
}

File: Chapter 04/Python/section_4_3_cross_cutting_concerns/after/order_service.py
from .logger import Logger

class OrderService:
    """
    2. THE CLASS "ASKS" FOR THE DEPENDENCY.
    """
    def __init__(self, logger: Logger):
        # The dependency is "injected" via the constructor!
        self.logger = logger

    def save_order(self, order):
        # 3. Use the abstraction (follows DIP)
        self.logger.log("Saving order...") # This uses the injected logger
        print("(AFTER_SERVICE) Order saved.")